library/core/src/hint.rs

```diff
@@ -310,13 +310,45 @@ pub fn spin_loop() {
 /// behavior in the calling code. This property makes `black_box` useful for writing code in which
 /// certain optimizations are not desired, such as benchmarks.
 ///
+/// In practice, `black_box` serves two purposes:
+///
+/// 1. It forces the input to be calculated, even if its results are never used
+/// 2. It prevents the compiler from making optimizations related to the value of the returned
+///    type
+///
+/// Note that `black_box` does not prevent its inputs from being optimized before they are passed
+/// to the function, though.
+///
+/// ```
+/// # use std::hint::black_box;
+/// #
+/// // This...
+/// let y = black_box(5 * 10);
+/// // ...will still be optimized to this:
+/// let y = black_box(50);
```

Comment:
This is a somewhat confusing way to explain it, because it talks only about optimizations, but the real issue has nothing to do with the optimizer *or* `black_box`, it is about Rust's semantics.

Functionally, `5 * 10` is *not* the input into `black_box` in `black_box(5 * 10)`. It is an **expression**, and as Rust is a strict language, it is then evaluated immediately. The compiler isn't really "replacing" things, it's implementing the language's eager evaluation semantics. Those semantics make these let-bindings functionally equivalent in this case:
```rust
let x = black_box(5 * 10);
// this evaluation of `x` follows the same steps as this evaluation of `y`:
let _0 = 5 * 10;
let y = black_box(_0);
```

The optimization that happens is merely downstream of the fact that there is no `black_box` around the "arguments" to `_0`.

Reply:
...I mean, you can talk about the compiler replacing things in the sense of "compilation is basically graph rewriting", but the point is that it's not an arbitrary motion of the optimizer, and what is more important is to express the underlying principle.

Reply:
Ok, let me revise my explanation with the example you gave above. I'm going to avoid talking about graphs, since that gets a little _too_ close to compiler internals, but I'll try my best to distill the rest!

Reply:
Ok, I took at shot at this in b6b374ce4d622a03a0342207d8a32ee726eab589. Tell me what you think!